### PR TITLE
CORDA-1892 CRaSH shell flow start fix for similar flow names.

### PR DESCRIPTION
**Pull request by josecoll**
If a node has two flows, where one's name is a longer version of the other's, they cannot be started

### CODE REVIEW

File: tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt [hunk 1]
```diff
@@ -249,23 +252,27 @@ object InteractiveShell {
...
             matches.forEachIndexed { i, s -> output.println("${i + 1}. $s", Color.yellow) }
             return
         }
 
         val flowClazz: Class<FlowLogic<*>> = if (classLoader != null) {
-            uncheckedCast(Class.forName(matches.single(), true, classLoader))
+            uncheckedCast(Class.forName(flowName, true, classLoader))
         } else {
-            uncheckedCast(Class.forName(matches.single()))
+            uncheckedCast(Class.forName(flowName))
```
Comment (chrisr3) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]: My poor little brain says that `flowName` _might_ be `null` here if `matches.size == 1`.
I.e. suppose `nameFragment = 'Burble'`, and the only flow were called `BurbleFlow`. Wouldn't this be a NPE?

Comment (chrisr3) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]:
```kotlin
val flowClazz: Class<FlowLogic<*>> = (flowName ?: matches[0]).let { className ->
    if (classLoader != null) {
        uncheckedCast(Class.forName(className, true, classLoader))
    } else {
        uncheckedCast(Class.forName(className))
    }
}
```

File: tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt [hunk 2]
```diff
@@ -255,11 +255,13 @@ object InteractiveShell {
...
-        val flowName = matches.find { it.endsWith(nameFragment) }
-        if (matches.size > 1 && flowName == null) {
+        val flowName = matches.find { it.endsWith(nameFragment) } ?: if (matches.size > 1) {
             output.println("Ambiguous name provided, please be more specific. Your options are:")
             matches.forEachIndexed { i, s -> output.println("${i + 1}. $s", Color.yellow) }
             return
+        } else {
+            output.println("No matching flow found, run 'flow list' to see your options.", Color.red)
+            return
         }
```
Comment (chrisr3) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]: But in your test case, there _is_ a flow called `BurbleFlow` so why shouldn't `nameFragment = 'Burble'` match?
Wouldn't this make more sense?
```kotlin
val flowName = matches.find { it.endsWith(nameFragment) } ?: if (matches.size > 1) {
   // ambiguous
} else if (matches.isEmpty()) {
   // no flow found
} else {
    // Unique match - use this one!
    matches[0]
}
```

Comment (josecoll) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]: The code will only execute a flow whereby the flowName is completely specified (either as a single classname or FQ package name) and is not ambiguous.
So in your example, "Burble" is not a valid flow, but "BurbleFlow" and "net.corda.tools.shell.BurbleFlow" are.
Your interpretation is a different requirement and not related to this fix.

Comment (josecoll) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]: Also note that `matches.find { it.endsWith(nameFragment) }` means that `Burble` will never match as no actual flows end in `Burble` ;-)

Comment (chrisr3) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]: So is this a change in behaviour? Would "Burble" previously have matched `BurbleFlow`?

**Comment by josecoll** [comment]
@chrisr3 reverted back to original behaviour as per your latest comment.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt (hunk 2 above)
Comment (josecoll) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]: Good point - let me ensure the original behaviour is unchanged.

**Pull request merged**